src/hotspot/share/runtime/handshake.cpp

```diff
@@ -786,6 +792,17 @@ bool HandshakeState::suspend() {
   } else {
     SuspendThreadHandshake st;
     Handshake::execute(&st, _handshakee);
+
+    if (st.did_suspend() && _handshakee->is_virtual()) {
+      // Thread suspension works under JvmtiVTMSTransitionDisabler protection, so we need to wait
+      // for virtual thread to reach a safe state before the JVMTI SuspendThread is returned.
+      while (_handshakee->thread_state() != _thread_blocked &&
+             _handshakee->thread_state() != _thread_in_native) {
+        ThreadBlockInVM tbivm(self);
+        MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+        _lock.wait_without_safepoint_check(1);
+      }
```

Comment:
This would normally be incorrectly coded as you do not hold the lock around the state-change and so you may miss the notification. However, it is possible in this case that the overall handshake protocol prevents that from happening, but I cannot easily determine that.